### PR TITLE
feat: CFD-298 Fare day end global settings

### PR DESCRIPTION
**Pull request by jacsamell**
# Description

-   Fare day end global settings

# Testing instructions

-   Check that you can create / edit the Fare Day End setting in Global settings
-   Check the additional GS pages show the Fare Day End info correctly

# Type of change

-   [X] feat - A new feature
-   [ ] fix - A bug fix
-   [ ] docs - Documentation only changes
-   [ ] codestyle - Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
-   [ ] refactor - A code change that neither fixes a bug nor adds a feature
-   [ ] perf - A code change that improves performance
-   [ ] test - Adding missing tests or correcting existing tests
-   [ ] build - Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
-   [ ] ci - Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)
-   [ ] chore - Other changes that don't modify src or test files
-   [ ] revert - Reverts a previous commit
-   [ ] content - Changes to content or copy

### CODE REVIEW

File: repos/fdbt-site/src/pages/manageFareDayEnd.tsx [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                        <ErrorSummary errors={errors} />
+
+                        <h1 className="govuk-heading-xl">Fare day end</h1>
+                        <p className="govuk-body govuk-!-margin-bottom-4" id={'fare-day-text'}>
+                            If your fare day extends past midnight, enter the time on the following morning when your
+                            tickets expire
+                        </p>
+
+                        <div className="govuk-inset-text">
+                            Enter times in 24hr format. For example 0900 is 9am, 1730 is 5:30pm.
```
Comment (RobCatton) [inline repos/fdbt-site/src/pages/manageFareDayEnd.tsx:49]: I think this should say "Enter the time in 24hr format. For example 0900 is 9am, 1730 is 5:30pm."

File: repos/fdbt-site/src/pages/manageFareDayEnd.tsx [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+                        </div>
+
+                        <CsrfForm action="/api/manageFareDayEnd" method="post" csrfToken={csrfToken}>
+                            <FormGroupWrapper errorIds={[fareDayEndInputId]} errors={errors}>
+                                <fieldset className="govuk-fieldset">
+                                    <FormElementWrapper
+                                        errors={errors}
+                                        errorId={fareDayEndInputId}
+                                        errorClass="govuk-input--error"
```
Comment (RobCatton) [inline repos/fdbt-site/src/pages/manageFareDayEnd.tsx:58]: Using this page, it wasn't very obvious to me that I would be clearing a saved fare day end time if I clicked save after emptying the text box. Its also a little strange that upon first load, and when adding one, if I click save when not adding anything, it says it has saved. I think we either need to have some other functionality (clear button?) or some hint text to explain that they can clear it by emptying the box.

File: repos/fdbt-site/src/pages/globalSettings.tsx [hunk 3]
```diff
@@ -56,6 +57,12 @@ const GlobalSettings = ({ globalSettingsCounts, referer }: GlobalSettingsProps):
                             description="Define certain days and time periods that your tickets can be used within"
                             count={globalSettingsCounts.timeRestrictionsCount}
                         />
+                        <SettingOverview
+                            href="/manageFareDayEnd"
+                            name="Fare day end"
+                            description="Define the time at which your fare day ends"
+                            count={globalSettingsCounts.fareDayEndSet}
```
Comment (RobCatton) [inline repos/fdbt-site/src/pages/globalSettings.tsx:64]: When the count is zero here, it has nothing displayed, which looks a little odd given we have zero's above. I think something like "Not specified" or "Uncustomised" would be good to have - worth checking with Esmond.

File: repos/fdbt-site/src/pages/api/manageFareDayEnd.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import { updateSessionAttribute } from '../../utils/sessions';
+import { NextApiRequestWithSession } from '../../interfaces';
+import { GS_FARE_DAY_END_ATTRIBUTE } from '../../constants/attributes';
+import { removeExcessWhiteSpace, isValid24hrTimeFormat } from './apiUtils/validator';
+import { upsertFareDayEnd } from '../../data/auroradb';
+import { fareDayEndInputId } from '../manageFareDayEnd';
+
+export default async (req: NextApiRequestWithSession, res: NextApiResponse): Promise<void> => {
+    try {
+        const fareDayEnd: string | undefined = removeExcessWhiteSpace(req.body.fareDayEnd);
```
Comment (RobCatton) [inline repos/fdbt-site/src/pages/api/manageFareDayEnd.ts:12]: This function always returns a string, do we need to have the " | undefined " part?

File: repos/fdbt-site/src/pages/manageFareDayEnd.tsx (hunk 2 above)
Comment (jacsamell) [reply repos/fdbt-site/src/pages/manageFareDayEnd.tsx:58]: Good shout, will discuss in standup tomorrow 👍

File: repos/fdbt-site/src/pages/globalSettings.tsx (hunk 3 above)
Comment (jacsamell) [reply repos/fdbt-site/src/pages/globalSettings.tsx:64]: Will discuss tomorrow 👍

File: repos/fdbt-site/src/pages/api/manageFareDayEnd.ts (hunk 4 above)
Comment (jacsamell) [reply repos/fdbt-site/src/pages/api/manageFareDayEnd.ts:12]: 👍  will remove

File: repos/fdbt-site/src/pages/manageFareDayEnd.tsx (hunk 1 above)
Comment (jacsamell) [reply repos/fdbt-site/src/pages/manageFareDayEnd.tsx:49]: Good shout 👍

Comment (jacsamell) [reply repos/fdbt-site/src/pages/manageFareDayEnd.tsx:49]: Done

File: repos/fdbt-site/src/pages/manageFareDayEnd.tsx (hunk 2 above)
Comment (jacsamell) [reply repos/fdbt-site/src/pages/manageFareDayEnd.tsx:58]: Done

File: repos/fdbt-site/src/pages/globalSettings.tsx (hunk 3 above)
Comment (jacsamell) [reply repos/fdbt-site/src/pages/globalSettings.tsx:64]: Updated text

File: repos/fdbt-site/src/pages/api/manageFareDayEnd.ts (hunk 4 above)
Comment (jacsamell) [reply repos/fdbt-site/src/pages/api/manageFareDayEnd.ts:12]: Done

**Review by RobCatton — APPROVED** [review]

**Pull request merged**